Generic LDNodes and RO-Crate parsing rework

HLWeil opened the pull request:
Summary
- Coming soon ™️ 

To-do:
- [ ] DOI and PubmedID as identifier
- [ ] Bioschemas context
- [ ] Some `@id` are empty (LabProtocol)
- [ ] Top-level roundabout tests
- [ ] Link datafiles instead of fragment selectors from datasets

HLWeil commented:
@floWetzels could you also take a look onto whether #326 is actually fixed or not (e.g. by inspecting a written RO-Crate metadata json)

floWetzels commented:
I tested the metadata json file generated for the ArcPrototype with the rocrate-validator. I found the following errors that need to be fixed:
- Studies und assays have wrong ids - the path to the isa-xlsx file. It should be the path to the directory ending with a `/`.
- Studies and assays are missing the `hasPart` property
- The root data entity has no `license`
- The root data entity has no `datePublished`

Furthermore,  I noticed the following when parsing the file manually:
- Paths must be relative to the root of the ro-crate
- Ad-hoc ids (e.g. for processes, paramters, parameter values) should begin with a `#` symbol
- The id of PropertyValue objects looks weird to me. It contains the `propertyID`, which could be a potential source for errors. Previously, it was generated only from `name` and `value`, if I am not mistaken.

HLWeil commented:
I incorporated your suggested fixes and included the test file in [d3d786e](https://github.com/nfdi4plants/ARCtrl/pull/496/commits/d3d786e6a4f8a6c5207a9e339043471e4f84d650)

kMutagene commented:
@HLWeil i think adding a test that performs the validation of the ro-crate file that @floWetzels did manually is needed. If not in this PR then via separate issue

kMutagene commented on src/ROCrate/Generic/File.fs:
```diff
@@ -0,0 +1,123 @@
+    static member genId(name : string) =
+        $"{name}"
+
+    static member validate(dt : LDNode, ?context : LDContext) =
```
Is this the equivalent of the missing file profile for https://github.com/nfdi4plants/isa-ro-crate-profile/issues/57?

HLWeil replied:
Yes

HLWeil replied:
... I mean yesn't. 

It isn't missing, File = [Data](https://github.com/nfdi4plants/isa-ro-crate-profile/blob/release/profile/isa_ro_crate.md#data)

kMutagene commented on src/ROCrate/Generic/Organization.fs:
Is this the missing organization profile for https://github.com/nfdi4plants/isa-ro-crate-profile/issues/57 ?

HLWeil replied:
Yes

kMutagene commented on src/ROCrate/Generic/PostalAddress.fs:
Is this the missing postaladdress profile for https://github.com/nfdi4plants/isa-ro-crate-profile/issues/57 ? If we have 0 mandatory fields, maybe we just don't need to support this type?

HLWeil replied:
Well there were discussions about it for the specification. Also requests for the implementation: https://github.com/nfdi4plants/ARCtrl/issues/465

I currently handle it in ARC Scaffold by just writing the json blob in the string field.

kMutagene commented on src/ROCrate/Generic/PropertyValue.fs:
```diff
@@ -0,0 +1,171 @@
+    static member setValueReference(pv : LDNode, valueReference : string, ?context : LDContext) =
+        pv.SetProperty(PropertyValue.valueReference, valueReference, ?context = context)
+
+    static member validate(pv : LDNode, ?context : LDContext) =
```
Is this missing validate functions for PropertyValue - DOI and PropertyValue - PubMedID ?

HLWeil replied:
There's currently not a validate function for doi and pubmedID.

The convertes can be found here:
https://github.com/nfdi4plants/ARCtrl/blob/b755341a2b50cc7e9243c95e55ed423e6e63cfa8/src/ARCtrl/Conversion.fs#L1015

kMutagene commented on src/ROCrate/Generic/ScholarlyArticle.fs:
```diff
@@ -0,0 +1,103 @@
+    static member setAuthors(s : LDNode, authors : ResizeArray<LDNode>, ?context : LDContext) =
+        s.SetProperty(ScholarlyArticle.author, authors, ?context = context)
+
+    static member tryGetUrl(s : LDNode, ?context : LDContext) =
```
We dropped url as property of this type

kMutagene commented on src/ROCrate/ISAProfile/Assay.fs:
Why are you deleting all these types?

HLWeil replied:
That was a misunderstanding between me, myself and I

kMutagene commented on tests/ROCrate/ISAProfile/Person.Tests.fs:
```diff
@@ -14,7 +14,7 @@ let mandatory_properties = Person(
 let all_properties = Person(
```
How are these tests still passing when you deleted all the types tested? maybe not seeing the forest for the trees here

kMutagene commented on tests/ROCrate/Main.fs:
```diff
@@ -3,18 +3,19 @@ module ARCtrl.ROCrate.Tests
-    Tests.ScholarlyArticle.main
+    Tests.LDContext.main
+    Tests.LDNode.main
+    //Tests.Dataset.main
```
hmm i think i found the issue, you are not running the tests. Why fix them then?

kMutagene reviewed (changes requested):
See comments

HLWeil commented:
Alright the comments should be resolved. 
- I reincluded the initial ROCrate Types, and renamed the static classes using the `LD` prefix. 
- LDPropertyValue now has DOI and PubMedID helper functions
- Implemented and tested DOI and PubMedID backwards compatability (using `url` and `sameAs`)
